### ISSUE TITLE
Weird warnings when trying to freeze some pipeline components

### ISSUE THREAD

**Issue by marc-poweruser** [explosion/spaCy#7568]
# How to reproduce the behaviour

I'm getting a weird set of warnings when trying to freeze components of my pipeline.  In this simplified test case, I've got two transformers, a tagger, and a parser.  I'm trying to freeze the tagger and train the parser, with the tagger listening to one (frozen) transformer and the parser listening to the other (unfrozen) transformer.  The tagger and its transformer are both being sourced from the en_core_web_trf model.  The tagger is listening to a transformer named "transformer" since that's its name in en_core_web_trf, while the parser is listening to a transformer named "transformer2".

The warnings include the following (full text of warnings follows below):
1. Component 'parser' will be (re)trained, but it needs the component 'transformer' which is frozen.
2. Component 'transformer2' will be (re)trained, but the component 'tagger' depends on it via a listener and is frozen.
3. Component 'ner' will be (re)trained, but it needs the component 'transformer' which is frozen.

These seem weird to me because:
1. The parser should be listening upstream to "transformer2" not "transformer"
2. The tagger should be listening upstream to "transformer" not "transformer2"
3. ner isn't a component in my pipeline.  Is it because it's a component of the model source tagger and one of my transformers is using?  If so, how do I deal with that appropriately?

It's as if the parser and tagger components are each listening to both transformers for some reason.

The warnings suggest using replace_listeners, so I tried adding "replace_listeners = ["model.tok2vec"]" to the tagger, but adding it results in the error "TypeError: 'FullTransformerBatch' object is not iterable".  I have no idea how to add it to the parser since it's referencing a factory not a source, which I thought didn't work with replace_listeners.  I don't understand why I'd need replace_listeners for either anyway since they should be listening to different transformers to begin with, so copying the transformer listener shouldn't matter, right?

So many questions.  Anyway, here are the particulars - I'd appreciate any assistance if I'm doing something wrong.

# Your Environment
- LINUX
- spaCy 3.0
- Python 3.6
- Pytorch 1.6
- GPU
- 16MB RAM

# Config
Here's my config file:

```
[paths]
train = null
dev = null
vectors = null
init_tok2vec = null
model = "../models-v3"

[system]
gpu_allocator = "pytorch"
seed = 0

[nlp]
lang = "en"
pipeline = ["transformer","transformer2","tagger","parser"]
batch_size = 128
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.parser]
factory = "parser"
learn_tokens = false
min_action_freq = 30
moves = null
update_with_oracle_cut_size = 100

[components.parser.model]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "parser"
extra_state_tokens = false
hidden_width = 128
maxout_pieces = 3
use_upper = false
nO = null

[components.parser.model.tok2vec]
@architectures = "spacy-transformers.TransformerListener.v1"
grad_factor = 1.0
pooling = {"@layers":"reduce_mean.v1"}
upstream = "transformer2"

[components.tagger]
source = "en_core_web_trf"

[components.tagger.model]

[components.tagger.model.tok2vec]
@architectures = "spacy-transformers.TransformerListener.v1"
grad_factor = 1.0
pooling = {"@layers":"reduce_mean.v1"}
upstream = "transformer"

[components.transformer]
source = "en_core_web_trf"

[components.transformer2]
factory = "transformer"
name = "transformer2"
max_batch_items = 4096
set_extra_annotations = {"@annotation_setters":"spacy-transformers.null_annotation_setter.v1"}

[components.transformer2.model]
@architectures = "spacy-transformers.TransformerModel.v1"
name = "roberta-base"

[components.transformer2.model.get_spans]
@span_getters = "spacy-transformers.doc_spans.v1"

[components.transformer2.model.tokenizer_config]
use_fast = true

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 500
gold_preproc = false
limit = 0
augmenter = null

[training]
accumulate_gradient = 3
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
patience = 1600
max_epochs = 2
max_steps = 20000
eval_frequency = 200
frozen_components = ["transformer","tagger"]
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_padded.v1"
discard_oversize = true
size = 2000
buffer = 256
get_length = null

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001

[training.optimizer.learn_rate]
@schedules = "warmup_linear.v1"
warmup_steps = 250
total_steps = 20000
initial_rate = 0.00005

[training.score_weights]
dep_las_per_type = null
sents_p = null
sents_r = null
ents_per_type = null
cats_score_desc = null
cats_micro_p = null
cats_micro_r = null
cats_micro_f = null
cats_macro_p = null
cats_macro_r = null
cats_macro_f = null
cats_macro_auc = null
cats_f_per_type = null
cats_macro_auc_per_type = null
tag_acc = 0.25
dep_uas = 0.12
dep_las = 0.12
sents_f = 0.0
ents_f = 0.25
ents_p = 0.0
ents_r = 0.0
cats_score = 0.25

[pretraining]

[initialize]
vectors = null
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]
```

# Warnings

Here is the full text of the warnings I'm getting:

```
[2021-03-25 22:01:50,557] [WARNING] [W086] Component 'parser' will be (re)trained, but it needs the component 'transformer' which is frozen. You can either freeze both, or neither of the two. If you're sourcing the component from an existing pipeline, you can use the `replace_listeners` setting in the config block to replace its token-to-vector listener with a copy and make it independent. For example, `replace_listeners = ["model.tok2vec"]` See the documentation for details: https://spacy.io/usage/training#config-components-listeners
WARNING:spacy:[W086] Component 'parser' will be (re)trained, but it needs the component 'transformer' which is frozen. You can either freeze both, or neither of the two. If you're sourcing the component from an existing pipeline, you can use the `replace_listeners` setting in the config block to replace its token-to-vector listener with a copy and make it independent. For example, `replace_listeners = ["model.tok2vec"]` See the documentation for details: https://spacy.io/usage/training#config-components-listeners
[2021-03-25 22:01:50,557] [WARNING] [W086] Component 'ner' will be (re)trained, but it needs the component 'transformer' which is frozen. You can either freeze both, or neither of the two. If you're sourcing the component from an existing pipeline, you can use the `replace_listeners` setting in the config block to replace its token-to-vector listener with a copy and make it independent. For example, `replace_listeners = ["model.tok2vec"]` See the documentation for details: https://spacy.io/usage/training#config-components-listeners
WARNING:spacy:[W086] Component 'ner' will be (re)trained, but it needs the component 'transformer' which is frozen. You can either freeze both, or neither of the two. If you're sourcing the component from an existing pipeline, you can use the `replace_listeners` setting in the config block to replace its token-to-vector listener with a copy and make it independent. For example, `replace_listeners = ["model.tok2vec"]` See the documentation for details: https://spacy.io/usage/training#config-components-listeners
[2021-03-25 22:01:50,558] [WARNING] [W087] Component 'transformer2' will be (re)trained, but the component 'tagger' depends on it via a listener and is frozen. This means that the performance of 'tagger' will be degraded. You can either freeze both, or neither of the two. If you're sourcing the component from an existing pipeline, you can use the `replace_listeners` setting in the config block to replace its token-to-vector listener with a copy and make it independent. For example, `replace_listeners = ["model.tok2vec"]` See the documentation for details: https://spacy.io/usage/training#config-components-listeners
WARNING:spacy:[W087] Component 'transformer2' will be (re)trained, but the component 'tagger' depends on it via a listener and is frozen. This means that the performance of 'tagger' will be degraded. You can either freeze both, or neither of the two. If you're sourcing the component from an existing pipeline, you can use the `replace_listeners` setting in the config block to replace its token-to-vector listener with a copy and make it independent. For example, `replace_listeners = ["model.tok2vec"]` See the documentation for details: https://spacy.io/usage/training#config-components-listeners
```

**Comment by polm** [comment]
I need to look at this in more detail, but at first glance it looks like the same issue as https://github.com/explosion/spaCy/discussions/7463. There was a bug where you would get warnings for listeners on components even if they weren't in your pipeline; in general you should be able to ignore them and nothing in particular should happen. Sorry for the confusion. 

Also, please do use `python -m spacy info` to post your environment details - for detailed issues like this full version number of everything is important.

**Comment by marc-poweruser** [comment]
Thanks, will do - here are the environment detail from spacy info:

spaCy version    3.0.5                         
Location         /opt/conda/lib/python3.6/site-packages/spacy
Platform         Linux-4.14.219-164.354.amzn2.x86_64-x86_64-with-debian-buster-sid
Python version   3.6.13                        
Pipelines        en_core_web_trf (3.0.0)

**Comment by svlandeg** [comment]
Hi!

I had a look into the issues you're running into, and there are indeed a few bugs here.

> 3. Component 'ner' will be (re)trained, but it needs the component 'transformer' which is frozen.

This issue is [resolved](https://github.com/explosion/spaCy/pull/7471) in the current version on `master` as pointed out by Paul.

> 1. Component 'parser' will be (re)trained, but it needs the component 'transformer' which is frozen.
> 2.  Component 'transformer2' will be (re)trained, but the component 'tagger' depends on it via a listener and is frozen.
    
Both of these are caused by the fact that in the original `en_core_web_trf` config file, both the parser and the tagger listen to 
```
upstream = "*"
```
which basically matches each Transformer earlier in the pipeline.

This is problematic, because there's no easy way I can think of right now to override this. Your config file has the following:
```
[components.tagger]
source = "en_core_web_trf"

[components.tagger.model]

[components.tagger.model.tok2vec]
@architectures = "spacy-transformers.TransformerListener.v1"
grad_factor = 1.0
pooling = {"@layers":"reduce_mean.v1"}
upstream = "transformer"
```
But note that the entire `[components.tagger.model]` is obsolete and effectively ignored when initializing the pipeline, because the configuration from the sourced component from `en_core_web_trf` is pulled in. This also means that your note of only listening to 
```
upstream = "transformer"
```
is ignored :(

I'm trying to think what the best way to resolve this is, because the config you created should work, and you should be able to have multiple independent transformers in the pipeline. You froze your components correctly so there shouldn't be any warnings.

A very ugly fix, just so you can continue working is to adjust the original `config.cfg` of your installed `en_core_web_trf` model (either the original model in `site-packages`, or a copied version with a different name and source from that name). You'd change `upstream` to `transformer` for the tagger, and to `transformer2` for the parser. Do note that this will make that config file / model unusable because now it refers to a `transformer2` that is not yet in the pipeline! But running your config from your original post should work after that.

In the meantime, we should definitely think about how to resolve this use-case more properly.

**Comment by marc-poweruser** [comment]
The temporary fix worked - thank you!  And I appreciate the explanation - at least now I understand what was going on.

**Comment by marc-poweruser** [comment]
To be clear, I'm still getting warnings with the ner (at least in my slightly more complex real use case) but can solve them in the same way.  I mention it only because it sounded above like you expected the ner warning to have been fixed in the recent release.

**Comment by svlandeg** [comment]
Sorry, I should have been more clear. The warnings around the NER are resolved in a recent PR on `master`, but this hasn't made it in a release yet.

**Comment by marc-poweruser** [comment]
Ah, got it - thanks!

**Comment by govertbuijs** [comment]
On 3.0.6, the same error appears if the **en_core_web_trf** model is used in the ner_demo_update example.

**Comment by svlandeg** [comment]
Hi all,

Just as a note, we've been discussing this internally. There is no great way to fix the `"*"` behaviour in current models when you want to have multiple `transformer` components in the pipeline. You can resort to the hack I described, changing the config of the pretrained pipeline (you can make a local copy first, ofcourse).

However, we'll fix this for the models that we'll release for the next 3.1 version of spaCy. We'll replace the `"*"` with an actual name so it won't match any new `transformer` components you'd add to a custom pipeline.

**Comment by svlandeg** [comment]
This is fixed in 3.1 - for the old models the "hack" described above will work. Finally closing this - thanks for your patience!
